File: src/main/java/de/fossag/HackatronWrapper.java
package de.fossag;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.PrintWriter;
import java.net.Socket;

/**
 * Simple wrapper class for {@link IHackatronClient} implementations that handles
 * network and logging.
 */
public class HackatronWrapper {

  private String host;
  private int port;

  private IHackatronClient hackatronClient;

  public HackatronWrapper(String host, int port, IHackatronClient hackatronClient) {
    this.host = host;
    this.port = port;
    this.hackatronClient = hackatronClient;
  }

  public void run() throws IOException {
    System.out.println("Connecting to " + host + ":" + port);

    // Set up socket and wire up streams
    Socket socket = new Socket(host, port);
    OutputStream out = socket.getOutputStream();
    PrintWriter writer = new PrintWriter(out, true);
    InputStream in = socket.getInputStream();
    BufferedReader reader = new BufferedReader(new InputStreamReader(in));

    System.out.println("Connection successful");

    // The message sender logs the outgoing message and stuffs it into the output stream
    hackatronClient.setMessageSender(s -> {
      System.out.println("OUT MSG: " + s);
      writer.println(s);
    });

    while (true) {
      if (socket.isClosed()) {
        return;
      }
      String line = reader.readLine();
      if (line == null) {
        return;
      }
      System.out.println("IN MSG: " + line);
      hackatronClient.onMessage(line);
    }

  }
}

File: src/main/java/de/fossag/hackatron/HackatronWrapper.kt
package de.fossag.hackatron

import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.io.PrintWriter
import java.net.Socket

/**
 * Simple wrapper class for [IHackatronClient] implementations that handles
 * network and logging.
 */
class HackatronWrapper(private val host: String, private val port: Int, private val hackatronClient: IHackatronClient) {
    @Throws(IOException::class)
    fun run() {
        println("Connecting to $host:$port")

        // Set up socket and wire up streams
        val socket = Socket(host, port)
        val out = socket.getOutputStream()
        val writer = PrintWriter(out, true)
        val `in` = socket.getInputStream()
        val reader = BufferedReader(InputStreamReader(`in`))
        println("Connection successful")

        // The message sender logs the outgoing message and stuffs it into the output stream
        hackatronClient.setMessageSender { s: String ->
            Logger.log("OUT MSG: $s")
            writer.println(s)
        }
        while (true) {
            if (socket.isClosed) {
                return
            }
            val line = reader.readLine() ?: return
            Logger.log("IN MSG: $line")
            hackatronClient.onMessage(line)
        }
    }
}
